BeanstalkFarms/Beanstalk
# feat: [dex-ui]: Well Creator UI

- [x] Theme Changes + Routing + Well Implementation step UI: https://github.com/BeanstalkFarms/Beanstalk/pull/896
- [x] Pump + Well Function step  + tokens + Well details UI: https://github.com/BeanstalkFarms/Beanstalk/pull/902
- [x] Deploy Preview UI:  https://github.com/BeanstalkFarms/Beanstalk/pull/913
- [x] Contract integration: https://github.com/BeanstalkFarms/Beanstalk/pull/915
- [x] UI bug fixes (e.g., support mobile):   https://github.com/BeanstalkFarms/Beanstalk/pull/918
- [x] UI bug fixes R2: https://github.com/BeanstalkFarms/Beanstalk/pull/920
- [x] UI bug fixes R3: https://github.com/BeanstalkFarms/Beanstalk/pull/923
- [x] UI Bug Fixes post testing: https://github.com/BeanstalkFarms/Beanstalk/pull/925
- [x] Decouple price data from basin api: https://github.com/BeanstalkFarms/Beanstalk/pull/929
- [x] Final Bug Passthrough: https://github.com/BeanstalkFarms/Beanstalk/pull/930

-----

### Test Preparation ###

Run `yarn` to update dependencies
Rebuild the Wells-Sdk: `yarn workspace @beanstalk/sdk-wells build`
Run Dex-UI locally from the project root: `yarn workspace dex-ui dev`

The apps have a hard time communicating with Metamask on a local fork. If Metamask isn't working, I'd recommend using Rabby wallet & import one of the Ethereum testing wallets.

Ethereum Test Account
| type | value |
| ---- | ------ |
| Address | 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266 | 
| PrivateKey | 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 |

You'll have to reset nonce data every time you reset your local fork. To do this in Rabby Wallet:
1. Open your Rabby wallet extension
2. Click on 'More'
3. Scroll all the way down and click on 'Clear Pending'
4. Click on 'Also reset my local nonce data and signature record'
5. Click confirm

### Setup ###
1. Run your local fork and send yourself any ERC20 tokens.
2. Navigate to /build in the nav bar
3. Click the 'Well Deployer' link button
4. Go through steps to deploy a new well

If you need a script to send tokens, deploy new instances of Well.SOL, CP2 & CP1, let me know. I can send you mine.

-----
## Test Flow ##

Token addresses for testing
| token   | address                                                                                    |
| ------- | -------------------------------------------------------- |
| BEAN   | 0xBEA0000029AD1c77D3d5D23Ba2D8893dB9d1Efab    |
| USDC  | 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48      |
| USDT   | 0xdAC17F958D2ee523a2206206994597C13D831ec7     |
| DAI       | 0x6B175474E89094C44Da98b954EedeAC495271d0F    |

### Vanilla Flow ###
***Does the deploy work when we use all recommended components & valid ERC20 tokens?***
1. Select recommended well implementation
2. Select recommended well function (Constant Product 2)
3. BEAN for token1 & other ERC20 for token2
4. Select recommended pump (Multi Flow Pump)
5. Keep recommended names
6. Deploy well with no salt and no liquidity
7. assertions: 
   - am I on the newly deployed well page? 

### Seed Liquidity Flow ###
***Does the deploy work when we also seed liquidity?***
1. Follow steps for vanilla flow up until deploy
2. Type liquidity amounts
3. Select salt 
4. Deploy
5. Assertions: 
   - Am I on the newly deployed well page? 
   - Did my tokens get used? (Check via swap page)
   - Does the well have liquidity? 
   - Did I receive the LP? (/Wells --> my positions tab)

### Variants and Form Checks ###
- Well Implementation
   - Valid ETH address enforced when using custom well implementation? 
   - Data input enforced when using custom well implementation? 
   - During deploy, does it deploy with a custom well implementation? (ask for script if you don't have)
   - During deploy, does it throw an error if the well implementation is not valid? 

- Well Function
   - Same checks as Well implementation

- Well Tokens
   - Token input enforced by form? 
   - Token uniqueness enforced by form?
   - Valid ERC20 token address enforced?

- Pump
   - Same checks as Well implementation
   - Can I deploy a well w/ no pumps successfully?

- Name and symbol
   - Token name & symbol enforced? 
   - Token name & symbol uniqueness enforced? (cannot share name & symbol with other wells)

- Liquidity
   - Can I deploy a well with one sided liquidity? 
   - Allowances approval before deploy enforced? 

- Salt
   - Only integers >= 0 enforced? 
   - Salt value >= 1 when seeding liquidity enforced?

- Wallet
   - Wallet connection before deploying a well enforced?
